### PR TITLE
AirfRANS: grad-clip=1.5 + T_max=5 + WD=1e-2 — triple compound

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Compound the three best AirfRANS findings:
1. **grad-clip=1.5** → 0.00935 (PR #2737, 26.4% improvement over clip=1.0)
2. **T_max=5** → 0.01271 (PR #2732, 3.9% improvement, more cosine cycles)
3. **WD=1e-2** → 0.01323 (PR #2709, 6.8% improvement, regularization)

Each improvement was discovered independently. If they're orthogonal, the compound should push well below 0.009. T_max=5 at clip=1.5 creates even more frequent phase transitions with moderate gradient clipping.

## Instructions

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --epochs 999 \
  --wandb-group airfrans-triple-compound \
  --wandb-name "haku/airfrans-gc15-tmax5-wd1e2"
```

Report: best val_primary/surface_mse, epoch, grad-clip stats, phase transition timing.

## Baseline

- **Primary metric:** val_primary/surface_mse = **0.00935** (PR #2737, lr=7e-4+grad-clip=1.5+T_max=10)
- **W&B baseline run:** 7bdiqnmi
- **External target:** 0.0043 (~2.2x gap)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.5 --no-use-ema --enable-fourier --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT haku:

## Results

### val_primary/surface_mse: 0.01326 at epoch 37 — NEGATIVE RESULT (+41.8% vs baseline 0.00935)

| Metric | Baseline (PR #2737) | This Run | Delta |
|--------|---------------------|----------|-------|
| val_primary/surface_mse | **0.00935** (ep40) | 0.01326 (ep37) | +41.8% (worse) |
| full_val/surface_mse_p | 0.0373 | 0.0529 | +41.8% (worse) |
| full_val/volume_mse | — | 0.0807 | — |
| test_primary/surface_mse | — | 0.02178 | — |

**W&B run:** 6f8zw995
**Epochs:** 41 (hit SENPAI_MAX_EPOCHS=50 cap / 30-min timeout)
**Grad-clip stats:** 277-353 events/epoch (77-98% of batches clipped), mean grad norm 9.8-23.8

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --epochs 999 \
  --wandb-group airfrans-triple-compound \
  --wandb-name "haku/airfrans-gc15-tmax5-wd1e2" \
  --agent haku
```

### What happened

The triple compound did NOT beat the baseline. The three improvements are **not orthogonal** — compounding them produced a worse result than grad-clip=1.5 alone at T_max=10.

**Key observations:**

1. **T_max=5 creates excessive volatility with grad-clip=1.5.** The peak-to-trough swings are enormous: epoch 37 hit 0.013 but epoch 38 immediately rebounded to 0.052 (4x swing). The baseline with T_max=10 gives the model more time to settle into deep basins between restarts.

2. **Phase transition occurred earlier but shallower.** The model broke through 0.1 by epoch 19 (vs ~epoch 35-40 in baseline runs), but the trough depth at epoch 37 (0.013) never matched the baseline's 0.00935. The more frequent restarts (T_max=5) may be disrupting the fine-grained optimization needed to reach the deepest basins.

3. **WD=1e-2 + T_max=5 may over-regularize.** WD=1e-2 alone improved results at T_max=10 (PR #2709: 0.01323), but paired with T_max=5 the extra regularization + more frequent perturbation prevents the model from fully exploiting each cosine trough.

4. **Grad norms are well-controlled** — mean grad norm decreased from ~20 (early) to ~10 (late epochs), with clip events dropping from 98% to 77% of batches. The training is stable but the cycling is too aggressive.

5. **Trough envelope:** ep5 (0.224) → ep14 (0.144) → ep16 (0.128) → ep21 (0.038) → ep27 (0.025) → ep33 (0.024) → ep36 (0.023) → ep37 (0.013). Still descending but never reached baseline depth.

### Suggested follow-ups

- **T_max=10 + WD=1e-2 + grad-clip=1.5**: The pairwise compound most likely to beat baseline. T_max=10 is already validated with clip=1.5, and WD=1e-2 improved results independently at T_max=10. This is the more conservative compound.
- **T_max=7 + grad-clip=1.5**: Intermediate cycling between T_max=5 (too aggressive) and T_max=10 (baseline). May find the sweet spot.
- **Longer training at T_max=5**: The trough envelope was still descending — 80+ epochs might eventually beat 0.00935, but T_max=10 seems more efficient at finding deep basins per epoch.

**Comment by tcapelle** [comment]
ADVISOR: The triple compound (gc=1.5+T_max=5+WD=1e-2) achieved 0.013262 — worse than baseline 0.00935. The issue is T_max=5 at gc=1.5 creates excessive volatility with massive 4x peak-to-trough swings.

However, we just got a MAJOR breakthrough on PR #2727: 4L/256d + gc=1.0 + WD=1e-2 + T_max=5 achieved **0.007264** — a 22.3% improvement. The architecture scaling + golden config is the winning direction.

Your next experiment: Test gc=1.5 at the NEW winning architecture. This could compound both breakthroughs:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-gc15
```

CRITICAL: Use SENPAI_MAX_EPOCHS=9999 — the previous run hit the epoch cap at only 61 min.

**Comment by tcapelle** [comment]
STUDENT haku:

## Results — Run 2: 4L/256d + gc=1.5 + T_max=5 + WD=1e-2

### val_primary/surface_mse: 0.0240 at epoch 25 — NEGATIVE RESULT (vs baseline 0.00935, vs PR #2727 gc=1.0: 0.00726)

| Metric | PR #2727 (gc=1.0) | This Run (gc=1.5) | Baseline (3L/192d gc=1.5) |
|--------|--------------------|--------------------|---------------------------|
| val_primary/surface_mse | **0.00726** | 0.0240 | 0.00935 |
| test_primary/surface_mse | — | 0.0274 | — |
| full_val/volume_mse | — | 0.103 | — |
| Epochs completed | ~50+ | 25 | 40 |

**W&B run:** d04mw0dg
**Epochs:** 25 (30-min timeout, ~1.2 min/epoch for 4L/256d)
**Grad-clip stats:** 283-342 events/epoch (79-95% clipped), mean grad norm 13.6-18.8

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.5 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-4L256d-gc15 \
  --wandb-name "haku/airfrans-4L256d-gc15-tmax5-wd1e2" \
  --agent haku
```

### What happened

gc=1.5 at 4L/256d did NOT match gc=1.0's performance. Two factors:

1. **Epoch starvation:** The 4L/256d model takes ~1.2 min/epoch vs ~0.7 min/epoch for 3L/192d. With SENPAI_TIMEOUT_MINUTES=30, this run only completed 25 epochs vs the ~50+ epochs PR #2727 achieved. The model was clearly still improving at cutoff (trough envelope: ep11→0.089, ep15→0.065, ep17→0.040, ep20→0.033, ep25→0.024 — consistent descent with no plateau).

2. **gc=1.5 may be suboptimal for 4L/256d:** Even accounting for fewer epochs, the descent rate appears slower than gc=1.0. The larger model may benefit from tighter clipping (gc=1.0) that constrains the higher-dimensional gradient space more effectively. At 3L/192d, the looser clip (1.5) helped because the gradient landscape is simpler.

**Trough envelope comparison (this run vs Run 1 at 3L/192d):**
- Epoch 20: 0.033 (4L/256d) vs 0.055 (3L/192d) — 4L is deeper at same epoch count
- Epoch 25: 0.024 (4L/256d) vs 0.025 (3L/192d at ep27) — similar
- The 4L architecture IS better per-epoch, but needs more wall-clock time

### Suggested follow-ups

- **Longer timeout for 4L/256d:** The model was still rapidly improving at cutoff. A 60-min budget would yield ~50 epochs and likely push below 0.01.
- **gc=1.0 confirmed as better for 4L/256d:** PR #2727's gc=1.0 result (0.00726) is clearly superior. The larger model benefits from tighter gradient control.
- **3L/192d + gc=1.5 may still be competitive** if given similar epoch counts — per-epoch efficiency is similar but wall-clock throughput is 70% higher.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2743 because ADVISOR: Closing — both runs failed. Run 1 (3L/192d): 0.013262. Run 2 (4L/256d): 0.02403 after only 25 epochs (crashed from timeout). The gc=1.5+WD=1e-2 combination consistently fails — WD fights gc=1.5's gradient headroom. We have dedicated experiments testing gc=1.5 WITHOUT WD (edward #2785) and with lighter WD (nezuko #2784). Thank you for the thorough work haku..
